src/app/modules/auth/auth.controller.ts

```diff
@@ -0,0 +1,195 @@
+
+/**
+ * Precondition: AuthMiddlewares.validateDomain must precede this handler.
+ * @returns 200 regardless, assumed to have passed domain validation.
+ */
+export const handleCheckUser: RequestHandler = async (_req, res) => {
+  return res.sendStatus(StatusCodes.OK)
+}
+
+/**
+ * Precondition: AuthMiddlewares.validateDomain must precede this handler.
+ */
+export const handleLoginSendOtp: RequestHandler<
+  {},
+  {},
+  { email: string }
+> = async (req, res) => {
+  // Joi validation ensures existence.
+  const { email } = req.body
+  const requestIp = getRequestIp(req)
+  const logMeta = {
+    action: 'handleSendLoginOtp',
+    email,
+    ip: requestIp,
+  }
+
+  // Create OTP.
+  const [otpErr, otp] = await to(AuthService.createLoginOtp(email))
+
+  if (otpErr) {
+    logger.error({
+      message: 'Error generating OTP',
+      meta: logMeta,
+      error: otpErr,
+    })
+    return res
+      .status(StatusCodes.INTERNAL_SERVER_ERROR)
+      .send(
+        'Failed to send login OTP. Please try again later and if the problem persists, contact us.',
+      )
+  }
+
+  // Send OTP.
+  const [sendErr] = await to(
+    MailService.sendLoginOtp({
+      recipient: email,
+      otp: otp!,
+      ipAddress: requestIp,
+    }),
+  )
+  if (sendErr) {
+    logger.error({
+      message: 'Error mailing OTP',
+      meta: logMeta,
+      error: sendErr,
+    })
+
+    return res
+      .status(StatusCodes.INTERNAL_SERVER_ERROR)
+      .send(
+        'Error sending OTP. Please try again later and if the problem persists, contact us.',
+      )
+  }
+
+  // Successfully sent login otp.
+  logger.info({
+    message: 'Login OTP sent successfully',
+    meta: logMeta,
+  })
+
+  return res.status(StatusCodes.OK).send(`OTP sent to ${email}!`)
+}
+
+/**
+ * Precondition: AuthMiddlewares.validateDomain must precede this handler.
+ */
+export const handleLoginVerifyOtp: RequestHandler<
+  {},
+  {},
+  { email: string; otp: string }
+> = async (req, res) => {
+  // Joi validation ensures existence.
+  const { email, otp } = req.body
+
+  const logMeta = {
+    action: 'handleLoginVerifyOtp',
+    email,
+    ip: getRequestIp(req),
+  }
+
+  const [verifyErr] = await to(AuthService.verifyLoginOtp(otp, email))
+
+  if (verifyErr) {
+    logger.warn({
+      message:
+        verifyErr instanceof InvalidOtpError
+          ? 'Login OTP is invalid'
+          : 'Error occurred when trying to validate login OTP',
+      meta: logMeta,
+      error: verifyErr,
+    })
+
+    if (verifyErr instanceof InvalidOtpError) {
+      return res.status(verifyErr.status).send(verifyErr.message)
+    }
+
+    // Unknown error, return generic error response.
+    return res
+      .status(StatusCodes.INTERNAL_SERVER_ERROR)
+      .send(
+        'Failed to validate OTP. Please try again later and if the problem persists, contact us.',
+      )
+  }
+
+  // OTP is valid, proceed to login user.
+  try {
+    const agency = await AuthService.getAgencyWithEmail(email)
```

Comment:
here we see that the middleware pattern with `validateDomain` results in an additional network call to the database to fetch the agency object again. can we revisit our discussion again? because (1) adhering to a pattern at the cost of performance seems overly-dogmatic, and (2) the actual logging could be done by the controller, which will set the context in which any lookup failure occurs.

Reply:
We can discuss after syncup later; but I also agree. I've experimented with passing agency into locals like before in b6a48db; seems like the best way to save some db calls. Take a look.